shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/ShoppingCart.scala

```diff
@@ -0,0 +1,254 @@
+import akka.persistence.typed.scaladsl.ReplyEffect
+import com.lightbend.lagom.scaladsl.persistence.AggregateEvent
+import com.lightbend.lagom.scaladsl.persistence.AggregateEventShards
+import com.lightbend.lagom.scaladsl.persistence.AggregateEventTag
+import com.lightbend.lagom.scaladsl.persistence.AggregateEventTagger
+import com.lightbend.lagom.scaladsl.persistence.AkkaTaggerAdapter
+import com.lightbend.lagom.scaladsl.playjson.JsonSerializer
+import com.lightbend.lagom.scaladsl.playjson.JsonSerializerRegistry
+import play.api.libs.json.Format
+import play.api.libs.json._
+
+import scala.collection.immutable.Seq
+
+object ShoppingCart {
+
+  // SHOPPING CART COMMANDS
+
+  // This is a marker trait for shopping cart commands.
+  // We will serialize them using Akka's Jackson support that is able to deal with the replyTo field.
+  // (see application.conf).
+  // Keep in mind that when configuring it on application.conf you need to use the FQCN which is:
+  // com.example.shoppingcart.impl.ShoppingCart$CommandSerializable
+  // Note the "$".
+  trait CommandSerializable
+
+  sealed trait Command extends CommandSerializable
+
+  final case class AddItem(itemId: String, quantity: Int, replyTo: ActorRef[Confirmation]) extends Command
+
+  final case class RemoveItem(itemId: String, replyTo: ActorRef[Confirmation]) extends Command
+
+  final case class AdjustItemQuantity(itemId: String, quantity: Int, replyTo: ActorRef[Confirmation]) extends Command
+
+  final case class Checkout(replyTo: ActorRef[Confirmation]) extends Command
+
+  final case class Get(replyTo: ActorRef[Summary]) extends Command
+
+  // SHOPPING CART REPLIES
+  final case class Summary(items: Map[String, Int], checkedOut: Boolean)
+
+  sealed trait Confirmation
+
+  final case class Accepted(summary: Summary) extends Confirmation
+
+  final case class Rejected(reason: String) extends Confirmation
+
+  implicit val summaryFormat: Format[Summary]               = Json.format
+  implicit val confirmationAcceptedFormat: Format[Accepted] = Json.format
+  implicit val confirmationRejectedFormat: Format[Rejected] = Json.format
+  implicit val confirmationFormat: Format[Confirmation] = new Format[Confirmation] {
+    override def reads(json: JsValue): JsResult[Confirmation] = {
+      if ((json \ "reason").isDefined)
+        Json.fromJson[Rejected](json)
+      else
+        Json.fromJson[Accepted](json)
+    }
+
+    override def writes(o: Confirmation): JsValue = {
+      o match {
+        case acc: Accepted => Json.toJson(acc)
+        case rej: Rejected => Json.toJson(rej)
+      }
+    }
+  }
+
+  // SHOPPING CART EVENTS
+  sealed trait Event extends AggregateEvent[Event] {
+    override def aggregateTag: AggregateEventTagger[Event] = Event.Tag
+  }
+
+  object Event {
+    val Tag: AggregateEventShards[Event] = AggregateEventTag.sharded[Event](numShards = 10)
+  }
+
+  final case class ItemAdded(itemId: String, quantity: Int) extends Event
```

Comment:
I included the cartId (entityId) in the events in https://github.com/akka/akka-samples/pull/135
The reason is that I think that events should be self contained and the cartId is an important domain value that is needed by downstream consumers (read-side). They may even be published to other services.

Knowing the format (separator) of the PersistenceId it's possible to extract the entityId from the PersistenceId, but that is not a good solution.

Reply:
In Lagom, events don't need the explicitly include the entityId because read-side processors don't process the actual event but an enveloped message that already provides the entityId.

Reply:
> In Lagom, events don't need the explicitly include the entityId because read-side processors don't process the actual event but an enveloped message that already provides the entityId.

This comment is only true in Lagom's `scaladsl`. Java users of Lagom still need to add the `cartId` on every event.

Reply:
yes, but that probably comes from parsed PersistenceId, since you know the separator.
You don't have to change. My comment was mostly for info about the difference.

Do you include the entityId somehow if events are published to Kafka?

Reply:
> Do you include the entityId somehow if events are published to Kafka?

Lagom doesn't doesn't do that automatically, no.

Users that need the `cartId` on their Kafka messages will have to manually add the `cartId` on every event (if using Java) or extract the `cartId` from the envelope when mapping the event into a Kafka-ready message.

Reply:
ok, thanks